### PR TITLE
NET-539: storage node assignment synchronization

### PR DESCRIPTION
**Pull request by harbu**
## Changes in a nutshell
- Add `blockNumber` to NodeRegistry's  `EtherumStorageEvent` and to the result of method `getStoredStreamsOf`, allowing us to determine which blockchain block the event or result applies to.
- Use this information in broker's storage plugin to synchronize events and polls appropriately. 
- Deprecate storage assignment event stream listening (we use smart contract event listeners to achieve the same) and thus broker's configuration setting `storageConfig.streamrAddress`.

## In detail

The client changes are straightforward. In the case of `EtherumStorageEvent`, we dig up the `blockNumber` directly from the ethers library callback function argument. For method `getStoredStreamsOf`, we modify the GraphQL query to also retrieve the latest block number with the result. In addition to these changes, some minor fixes to dependency injection to avoid circular dependency hassles that came up.

Broker has more major changes. Firstly, the core data structure and procedure for managing state is encapsulated in class `SetMembershipSynchronizer`. It is generalized to solve the problem of set membership over discrete time given full state updates and patches. In our case, the "elements" of the set are effectively `StreamPartID`(s) and we use The Graph polling for full state updates and blockchain-based events for patches.

Secondly, two separate data sources have been extracted from `StoragePlugin`. Everything related to polling full state data is in class `StoragePoller`, while event-based data handling can be found under class `StorageEventListener`. What responsibilities are left in `StoragePlugin` are (a) converting stream ids to stream parts, (b) taking care of storage cluster membership checks w.r.t. stream parts, and (c) instantiating the `StoragePoller` and `StorageEventListener` and connecting them with `SetMembershipSynchronizer`. Effectively `StoragePlugin` orchestrates the whole process of storage node assignments.

### CODE REVIEW

File: packages/broker/src/plugins/storage/SetMembershipSynchronizer.ts [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        return {
+            added: setDifference(this.computedState, previousComputedState),
+            removed: setDifference(previousComputedState, this.computedState)
+        }
+    }
+
+    /**
+     * Ingest a patch update
+     * @param elements the set of elements relevant to this patch
+     * @param operation indicates whether this is an assignment or a removal
```
Comment (teogeb) [inline packages/broker/src/plugins/storage/SetMembershipSynchronizer.ts]:
```suggestion
     * @param operation indicates whether this is an addition or a removal
```

File: packages/broker/src/plugins/storage/SetMembershipSynchronizer.ts [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+/**
+ * Represents a set that evolves over (discrete) time. The set is updated in
+ * two distinct ways:
+ *      1. Full state updates that define the set at a given time step.
+ *      2. Patch updates that add or remove a subset at a given time step.
+ */
+export class SetMembershipSynchronizer<E extends string> {
+    private lastState: ReadonlySet<E> = new Set<E>()
```
Comment (teogeb) [inline packages/broker/src/plugins/storage/SetMembershipSynchronizer.ts]: The `lastState` field is actually not needed.

File: packages/broker/src/plugins/storage/StorageConfig.ts [hunk 3]
```diff
@@ -1,288 +1,108 @@
...
+ *
+ *  Event-based assignments are great for picking up on changes quickly.
+ *  However, there is a risk of not receiving updates due to, e.g. connectivity
+ *  issues. Therefore, if the real-time system fails, polling acts as a sort-of
+ *  backup system.
+ */
 export class StorageConfig {
+    private readonly listener: StorageConfigListener
+    private readonly synchronizer = new SetMembershipSynchronizer<StreamPartID>()
+    private readonly clusterId: string
```
Comment (teogeb) [inline packages/broker/src/plugins/storage/StorageConfig.ts]: The `clusterId` field is redundant, not needed outside of the constructor.

File: packages/broker/src/plugins/storage/SetMembershipSynchronizer.ts [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    getState(): ReadonlySet<E> {
+        return this.computedState
+    }
+
+    /**
+     * Ingest a full state update at a given sequence number
+     * @param elements the set of _all_ elements at this sequence number
+     * @param sequenceNo the sequence number for this state
+     */
+    ingestState(elements: Set<E>, sequenceNo: number): Diff<E> {
```
Comment (teogeb) [inline packages/broker/src/plugins/storage/SetMembershipSynchronizer.ts]: Rename to `ingestSnapshot`? Then we'd have three clearly different words: `patch`, `snapshot` and `state`.

File: packages/broker/src/plugins/storage/StoragePoller.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+const logger = new Logger(module)
+
+/**
+ * Polls full state of storage node assignments on an interval.
+ */
+export class StoragePoller {
+    private readonly clusterId: string
+    private readonly pollInterval: number
+    private readonly streamrClient: StreamrClient
+    private readonly handleResult: (streams: Stream[], block: number) => void
```
Comment (teogeb) [inline packages/broker/src/plugins/storage/StoragePoller.ts]: Rename e.g. to `onNewSnapshot`?

File: packages/broker/src/plugins/storage/StorageEventListener.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+const logger = new Logger(module)
+
+/**
+ * Hooks up to `StreamrClient#registerStorageEventListener` to learn about
+ * stream assignment and removal events in real-time.
+ */
+export class StorageEventListener {
+    private readonly clusterId: string
+    private readonly streamrClient: StreamrClient
+    private readonly handleEvent: (stream: Stream, type: 'added' | 'removed', block: number) => void
```
Comment (teogeb) [inline packages/broker/src/plugins/storage/StorageEventListener.ts]: Rename to `onEvent`?

File: packages/broker/src/plugins/storage/StoragePoller.ts [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        handleResult: (streams: Stream[], block: number) => void
+    ) {
+        this.clusterId = clusterId
+        this.pollInterval = pollInterval
+        this.streamrClient = streamrClient
+        this.handleResult = handleResult
+    }
+
+    async start(): Promise<void> {
+        const schedulePoll = async () => {
```
Comment (teogeb) [inline packages/broker/src/plugins/storage/StoragePoller.ts]: Could use `scheduleAtInterval` helper method from `src/helpers/scheduler.ts`

File: packages/broker/test/unit/plugins/storage/StorageConfig.test.ts [hunk 8]
```diff
@@ -1,146 +1,178 @@
...
+function makeStubStream(streamId: string): Stream {
+    return {
+        id: toStreamID(streamId),
+        partitions: PARTITION_COUNT_LOOKUP[streamId]
+    } as Stream
+}
+
+describe(StorageConfig, () => {
+    let getStoredStreamsOf: jest.Mock<Promise<{ streams: Stream[], blockNumber: number }>, [nodeAddress: string]>
+    let emitStorageAssignmentEvent: ((event: EthereumStorageEvent) => any) | undefined
```
Comment (teogeb) [inline packages/broker/test/unit/plugins/storage/StorageConfig.test.ts]: Rename to `storageEventListener` (as already in`StorageEventListener.test.ts`)

File: packages/broker/test/unit/plugins/storage/StorageEventListener.test.ts [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    let storageEventListener: ((event: EthereumStorageEvent) => any) | undefined
+    let handleEvent: jest.Mock<void, [stream: Stream, type: 'added' | 'removed', block: number]>
+    let listener: StorageEventListener
+
+    beforeEach(() => {
+        stubClient = {
+            async getStream() {
+                return {
+                    id: 'streamId',
+                    partitions: 0
```
Comment (teogeb) [inline packages/broker/test/unit/plugins/storage/StorageEventListener.test.ts]: The `partitions` should be > 0. 

(Maybe good to use a value > 1, as multi-partition situation is not the most typical situation and therefore worth checking. Not very important in this as we don't have any partition processing here.)

File: packages/client/src/Encrypt.ts [hunk 10]
```diff
@@ -15,7 +15,7 @@ export default class PublisherEncryption implements Stoppable {
 
     constructor(
         private streamEndpoints: StreamEndpointsCached,
-        private keyExchange: PublisherKeyExchange,
+        @inject(delay(() => PublisherKeyExchange)) private keyExchange: PublisherKeyExchange,
```
Comment (teogeb) [inline packages/client/src/Encrypt.ts:18]: Does this fix something related to this PR?

**Review by teogeb — COMMENTED** [review]
Great improvement 👍

File: packages/broker/src/plugins/storage/SetMembershipSynchronizer.ts (hunk 2 above)
Comment (harbu) [reply packages/broker/src/plugins/storage/SetMembershipSynchronizer.ts]: Removed.

File: packages/broker/src/plugins/storage/StorageConfig.ts (hunk 3 above)
Comment (harbu) [reply packages/broker/src/plugins/storage/StorageConfig.ts]: Removed

File: packages/broker/src/plugins/storage/SetMembershipSynchronizer.ts (hunk 4 above)
Comment (harbu) [reply packages/broker/src/plugins/storage/SetMembershipSynchronizer.ts]: Good name. Updated.

File: packages/broker/src/plugins/storage/StoragePoller.ts (hunk 5 above)
Comment (harbu) [reply packages/broker/src/plugins/storage/StoragePoller.ts]: Renamed

File: packages/broker/src/plugins/storage/StorageEventListener.ts (hunk 6 above)
Comment (harbu) [reply packages/broker/src/plugins/storage/StorageEventListener.ts]: Renamed

File: packages/broker/src/plugins/storage/StoragePoller.ts (hunk 7 above)
Comment (harbu) [reply packages/broker/src/plugins/storage/StoragePoller.ts]: fixed

File: packages/broker/test/unit/plugins/storage/StorageConfig.test.ts (hunk 8 above)
Comment (harbu) [reply packages/broker/test/unit/plugins/storage/StorageConfig.test.ts]: fix

File: packages/broker/test/unit/plugins/storage/StorageEventListener.test.ts (hunk 9 above)
Comment (harbu) [reply packages/broker/test/unit/plugins/storage/StorageEventListener.test.ts]: Fixed

File: packages/client/src/Encrypt.ts (hunk 10 above)
Comment (harbu) [reply packages/client/src/Encrypt.ts:18]: Yeah the import paths have changed as a result of these changes, thereby causing circular dependencies.

**Review by teogeb — APPROVED** [review]

**Pull request merged**